File: ImperatorToCK3/Source/Imperator/Countries/Countries.cpp
#include "Countries.h"
#include "Country.h"
#include "../Families/Families.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"
#include <ranges>



Imperator::Countries::Countries(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void Imperator::Countries::registerKeys() {
	registerRegex(commonItems::integerRegex, [this](const std::string& countryID, std::istream& theStream) {
		std::shared_ptr<Country> newCountry = countryFactory.getCountry(theStream, commonItems::stringToInteger<unsigned long long>(countryID));
		countries.emplace(newCountry->getID(), newCountry);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}


Imperator::CountriesBloc::CountriesBloc(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void Imperator::CountriesBloc::registerKeys() {
	registerKeyword("country_database", [this](std::istream& theStream) {
		countries = Countries(theStream);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}


void Imperator::Countries::linkFamilies(const Families& theFamilies) {
	auto counter = 0;
	std::set<unsigned long long> idsWithoutDefinition;
	const auto& families = theFamilies.getFamilies();
	for (const auto& country : countries | std::views::values) {
		if (!country->getFamilies().empty()) {
			std::map<unsigned long long, std::shared_ptr<Family>> newFamilies;
			for (const auto& familyID : country->getFamilies() | std::views::keys) {
				const auto& familyItr = families.find(familyID);
				if (familyItr != families.end()) {
					newFamilies.insert(std::pair(familyItr->first, familyItr->second));
					counter++;
				}
				else {
					idsWithoutDefinition.insert(familyID);
				}
			}
			country->setFamilies(newFamilies);
		}
	}

	std::string warningString = "Families without definition:";
	if (!idsWithoutDefinition.empty()) {
		for (auto id : idsWithoutDefinition) {
			warningString += " ";
			warningString += std::to_string(id);
			warningString += ",";
		}
		warningString = warningString.substr(0, warningString.size() - 1); //remove last comma
		Log(LogLevel::Debug) << warningString;
	}
	
	Log(LogLevel::Info) << "<> " << counter << " families linked to countries.";
}

File: ImperatorToCK3/Imperator/Countries/Countries.cs
using commonItems;
using System.Collections.Generic;
using System.Linq;

namespace ImperatorToCK3.Imperator.Countries {
	public class Countries : Dictionary<ulong, Country> {
		public Countries() { }
		public Countries(BufferedReader reader) {
			var parser = new Parser();
			RegisterKeys(parser);
			parser.ParseStream(reader);
		}
		private void RegisterKeys(Parser parser) {
			parser.RegisterRegex(CommonRegexes.Integer, (reader, countryId) => {
				var newCountry = Country.Parse(reader, ulong.Parse(countryId));
				Add(newCountry.Id, newCountry);
			});
			parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);
		}
		public void LinkFamilies(Families.Families families) {
			SortedSet<ulong> idsWithoutDefinition = new();
			var counter = Values.Sum(country => country.LinkFamilies(families, idsWithoutDefinition));

			if (idsWithoutDefinition.Count > 0) {
				Logger.Debug($"Families without definition: {string.Join(", ", idsWithoutDefinition)}");
			}

			Logger.Info($"{counter} families linked to countries.");
		}
		public static Countries ParseBloc(BufferedReader reader) {
			var blocParser = new Parser();
			Countries countries = new();
			blocParser.RegisterKeyword("country_database", reader =>
				countries = new Countries(reader)
			);
			blocParser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);

			blocParser.ParseStream(reader);
			blocParser.ClearRegisteredRules();
			Logger.Debug($"Ignored CountryCurrencies tokens: {string.Join(", ", CountryCurrencies.IgnoredTokens)}");
			Logger.Debug($"Ignored RulerTerm tokens: {string.Join(", ", RulerTerm.IgnoredTokens)}");
			Logger.Debug($"Ignored Country tokens: {string.Join(", ", Country.IgnoredTokens)}");
			return countries;
		}
	}
}
